### PR TITLE
TreeNode classes refactor

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description

- move all tree generalized index calculations to a separate `GIndexUtils` class 
- make tree traversal a part of `TreeNode` functionality (`TreeNode.iterate()` methods)
- make default implementations for all 3 `TreeNode.updated()` overloads which invoke each other. So for minimal functionality just a single one needs to implemented. 
- move `LeafNode` and `BranchNode` to dedicated files 
- minor unrelated fix to `ProfilingRun`

The part of #3287 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/GIndexUtil.java [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+   *   <li><code>gIdxGetChildIndex(0b1101, 2) == 0b11</code>
+   *   <li><code>gIdxGetChildIndex(0b1110, 2) == 0b10</code>
+   *   <li><code>gIdxGetChildIndex(0b1111, 3) == SELF_G_INDEX</code>
+   *   <li><code>gIdxGetChildIndex(0b11, 2)</code> call would be invalid cause node with index 0b11
+   *       is at depth 1
+   * </ul>
+   */
+  public static long gIdxGetRelativeGIndex(long generalizedIndex, int childDepth) {
+    checkGIndex(generalizedIndex);
+    assert childDepth < 64;
```
Comment (mbaxter) [inline ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/GIndexUtil.java]: Should we assert `childDepth >= 0` as well?

File: ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/TreeUpdates.java [hunk 2]
```diff
@@ -105,25 +116,52 @@ private TreeUpdates(List<Long> gIndexes, List<TreeNode> nodes, long prefix, int
...
             gIndexes.subList(0, insIdx), nodes.subList(0, insIdx), lPrefix, heightFromLeaf - 1),
-        new TreeUpdates(
+        TreeUpdates.create(
             gIndexes.subList(insIdx, gIndexes.size()),
             nodes.subList(insIdx, nodes.size()),
             rPrefix,
             heightFromLeaf - 1));
   }
 
+  public List<TreeUpdates> splitToDepth(int depth) {
```
Comment (mbaxter) [inline ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/TreeUpdates.java]: Looks like this is unused

**Review by mbaxter — APPROVED** [review]
LGTM

File: ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/TreeUpdates.java (hunk 2 above)
Comment (Nashatyrev) [reply ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/TreeUpdates.java]: Yep, removing it for now

**Pull request merged**